Minor optimization to save_on_cpu

rohan-varma opened the pull request:
Stack from [ghstack](https://github.com/ezyang/ghstack):
* __->__ #70218

Offload to CPU by using `tensor.to("cpu", non_blocking=True)` instead of allocating an empty tensor and then non-blocking copy to it.

Also, cache `torch.cuda.is_available()` in the constructor.

Overall this provides a 3-4% win on some workloads with large model + large activations that we are testing.

Differential Revision: [D33246189](https://our.internmc.facebook.com/intern/diff/D33246189/)

**NOTE FOR REVIEWERS**: This PR has internal Facebook specific changes or comments, please review them on [Phabricator](https://our.internmc.facebook.com/intern/diff/D33246189/)!

rohan-varma committed:
Minor optimization to save_on_cpu

Offload to CPU by using `tensor.to("cpu", non_blocking=True)` instead of allocating an empty tensor and then non-blocking copy to it.

Also, cache `torch.cuda.is_available()` in the constructor.

Overall this provides a 3-4% win on some workloads with large model + large activations that we are testing.

Differential Revision: [D33246189](https://our.internmc.facebook.com/intern/diff/D33246189/)

**NOTE FOR REVIEWERS**: This PR has internal Facebook specific changes or comments, please review them on [Phabricator](https://our.internmc.facebook.com/intern/diff/D33246189/)!

[ghstack-poisoned]

rohan-varma commented on torch/autograd/graph.py:
```diff
@@ -113,16 +113,16 @@ class save_on_cpu():
-                dtype=tensor.dtype,
-                layout=tensor.layout,
-                pin_memory=(torch.cuda.is_available() and not tensor.is_sparse))
-            packed.copy_(tensor)
```
@albanD Any idea why this approach was chosen for save_on_cpu as opposed to `tensor.to`? Also, is there a reason `non_blocking=True` argument is not passed into `copy_`?

albanD replied:
As mentioned in the other issue, this split means that we do 2 copies of the Tensor instead of one. So I don't think we should do this.

non_blocking is not useful when moving to the cpu as this operation is always blocking.

rohan-varma replied:
I see, I wonder why FairScale checkpoint https://github.com/facebookresearch/fairscale/blob/main/fairscale/nn/checkpoint/checkpoint_activations.py#L273 uses it then

rohan-varma replied:
@albanD I checked some of the copy kernel code: https://github.com/pytorch/pytorch/blob/master/aten/src/ATen/native/cuda/Copy.cu#L208-L219, It does look like we still call cudaMemcpyAsync when copying from GPU to CPU when non_blocking=True, so the argument does make some difference? Not sure if I'm missing something here.

albanD replied:
Well, this is a bug. Nothing on the CPU side ensures proper synchronization.
Running the following on colab does fail:
```python
import torch

N = 100000000

a = torch.ones(N, device="cuda") * 42

for i in range(100):
    a * 2

out = a.to("cpu", non_blocking=True)
t = out.select(0, -1).item()
print(t)


out = a.to("cpu", non_blocking=False)
t = out.select(0, -1).item()
print(t)
```
it prints:
```
0.0
42.0
```

cc @ngimel  this seems to have been here forever. With explicit cases to handle it. Am I missing something here or we should fix this quickly?

albanD replied:
Note that the doc explicitly mentions that this arg should *only* have an effet when copying CPU -> GPU when the given CPU Tensor is already in pinned memory. And it should have no effet in all other cases.

ngimel replied:
People were (reasonably) asking for a way to do non-blocking copy to cpu, so this was enabled in #46878. And yes, since they explicitly request non-blocking copy, it's their responsibility to ensure proper synchronization. Docs should be updated.

ngimel commented on torch/autograd/graph.py:
```diff
@@ -113,16 +113,16 @@ class save_on_cpu():
-                pin_memory=(torch.cuda.is_available() and not tensor.is_sparse))
-            packed.copy_(tensor)
+            packed = tensor.to("cpu", non_blocking=True)
+            if self.cuda_avail and not tensor.is_sparse:
```
you don't need to separately pin memory, it will be automatically pinned with non-blocking copy:
```
In [1]: import torch

In [2]: a=torch.randn(4, device="cuda")

In [3]: b=a.to("cpu", non_blocking=True)

In [4]: b.is_pinned()
Out[4]: True
```
However, it also means that this is an asynchronous call, and `packed` won't contain the necessary data by the time this function returns. It will be the responsibility of the caller to synchronize before use (and to make sure `packed` doesn't go out of scope before it is actually used again)

rohan-varma replied:
Thanks for the review @ngimel!

It looks like we could simply do something like 

```
packed = tensor.to("cpu", non_blocking=True)
```

which also avoids the 2 copies that @albanD mentioned as an issue. Although, the concern about change in semantics makes sense, and this would probably be a BC-breaking change in that aspect. Since it helps the performance on a few reproducible FSDP workloads, could we consider adding a flag to allow non-blocking copies? @albanD

rohan-varma replied:
Also I was wondering whether the nonblocking semantics will cause issues in this particular case of `save_on_cpu`. Based on my understanding of how this is implemented in autograd, `packed` tensor as opposed to the input argument tensor will be saved by autograd. Although when this tensor is reloaded in `unpack` by autograd, we move it back to GPU (also nonblocking) so as long as `save_on_cpu` was previously being used with appropriate synchronization, this change should not incur any issues.

rohan-varma committed:
Update on "Minor optimization to save_on_cpu"


Offload to CPU by using `tensor.to("cpu", non_blocking=True)` instead of allocating an empty tensor and then non-blocking copy to it.

Also, cache `torch.cuda.is_available()` in the constructor.

Overall this provides a 3-4% win on some workloads with large model + large activations that we are testing.

Differential Revision: [D33246189](https://our.internmc.facebook.com/intern/diff/D33246189/)

**NOTE FOR REVIEWERS**: This PR has internal Facebook specific changes or comments, please review them on [Phabricator](https://our.internmc.facebook.com/intern/diff/D33246189/)!

[ghstack-poisoned]

rohan-varma committed:
Update on "Minor optimization to save_on_cpu"


Offload to CPU by using `tensor.to("cpu", non_blocking=True)` instead of allocating an empty tensor and then non-blocking copy to it.

Also, cache `torch.cuda.is_available()` in the constructor.

Overall this provides a 3-4% win on some workloads with large model + large activations that we are testing.

Differential Revision: [D33246189](https://our.internmc.facebook.com/intern/diff/D33246189/)

**NOTE FOR REVIEWERS**: This PR has internal Facebook specific changes or comments, please review them on [Phabricator](https://our.internmc.facebook.com/intern/diff/D33246189/)!

[ghstack-poisoned]

albanD commented on torch/autograd/graph.py:
```diff
@@ -112,8 +117,12 @@ class save_on_cpu():
 
     """
-    def __init__(self, pin_memory=False):
+    def __init__(self, pin_memory=False, non_blocking_copy_to_cpu=False):
```
This is redundant with the `pin_memory` argument no? We should have only one of them.

ngimel replied:
> Also @ngimel is there any reason not to do the copies async here by default? The engine does restore the ambient stream so I can't see any way for this to go wrong?

Async cpu copies require synchronizations somewhere. Otherwise, you'll call `pack_to_cpu`, `unpack_from_cpu` (all asynchronous), return from those calls, cpu tensor goes out of scope, that memory is reused, by the time actual copy queued by `unpack_from_cpu` happens it contains garbage.

rohan-varma replied:
@albanD I'm guessing we need to keep the `pin_memory` arg for BC. 

Although if we only have `pin_memory`, there will not be a flag to control whether we move the tensor to cpu with `non_blocking=True` in `pack_to_cpu`. Although, it was mentioned that non-blocking copy to CPU creates an automatically pinned memory tensor, so maybe the code can simply be something like:

```
def pack_to_cpu(tensor):
    oacked = tensor.to("cpu", non_blocking=pin_memory)
    return (tensor.device, packed)
```

and we can keep only the pin_memory argument? 

Regarding the synchronization issues, my understanding is that the CPU tensor will not go out of scope as it is saved by autograd so it won't be deleted before `unpack_from_cpu` has executed.

albanD replied:
@rohan-varma it is not, but it will be cleared after that backward function is executed.
But at that time, the async copy to the GPU might not have happened yet so it is not valid to release this CPU Tensor.

@ngimel isn't this issue what https://github.com/pytorch/pytorch/blob/3c2ae2b47cc859847e7915a406a7484401cba35e/aten/src/ATen/native/cuda/Copy.cu#L227 is trying to avoid?

ngimel replied:
It would prevent memory deallocation, but it wouldn't prevent tensor being filled with something else:
```
In [16]: a=torch.empty(10, pin_memory=True); a.fill_(1); b=torch.empty(10, device="cuda"); torch.cuda._sleep(5000000000); b.copy_(a, non_blocking=True); a.fil
    ...: l_(5); print(b);
tensor([5., 5., 5., 5., 5., 5., 5., 5., 5., 5.], device='cuda:0')
```
Here, we would expect `b` to be full of ones, because that was `a` contents at the time the copy was submitted, but instead it's filled with `5`. 
My point is, everywhere there's async operation, pytorch doesn't (and cannot) know how to do the right thing, and its authors job to either do necessary synchronizations, or leave comments why unfavorable sequence of events will never happen. Just changing blocking copies into non-blocking is not the right thing to do (even though it may still have the correct behavior).

albanD reviewed (commented):
I don't think there is any way for the user to access this CPU Tensor so it would be ok in this case indeed.
We should cleanup the API and this is good to go!

Also @ngimel is there any reason not to do the copies async here by default? The engine does restore the ambient stream so I can't see any way for this to go wrong?

rohan-varma committed:
Update on "Minor optimization to save_on_cpu"


Offload to CPU by using `tensor.to("cpu", non_blocking=True)` instead of allocating an empty tensor and then non-blocking copy to it.

Also, cache `torch.cuda.is_available()` in the constructor.

Overall this provides a 3-4% win on some workloads with large model + large activations that we are testing.

Differential Revision: [D33246189](https://our.internmc.facebook.com/intern/diff/D33246189/)

**NOTE FOR REVIEWERS**: This PR has internal Facebook specific changes or comments, please review them on [Phabricator](https://our.internmc.facebook.com/intern/diff/D33246189/)!

[ghstack-poisoned]

rohan-varma committed:
Update on "Minor optimization to save_on_cpu"


Offload to CPU by using `tensor.to("cpu", non_blocking=True)` instead of allocating an empty tensor and then non-blocking copy to it.

Also, cache `torch.cuda.is_available()` in the constructor.

Overall this provides a 3-4% win on some workloads with large model + large activations that we are testing.

Differential Revision: [D33246189](https://our.internmc.facebook.com/intern/diff/D33246189/)

**NOTE FOR REVIEWERS**: This PR has internal Facebook specific changes or comments, please review them on [Phabricator](https://our.internmc.facebook.com/intern/diff/D33246189/)!

[ghstack-poisoned]

albanD commented on torch/autograd/graph.py:
```diff
@@ -114,15 +115,7 @@ class save_on_cpu():
-                layout=tensor.layout,
-                pin_memory=(torch.cuda.is_available() and not tensor.is_sparse))
-            packed.copy_(tensor)
+            packed = tensor.to("cpu", non_blocking=pin_memory)
```
I think we want a comment block above the init that explains why it is ok to do non-blocking both ways here.

Also we should have a scary comment here and in the unpack to make sure that no one will ever try to use the "packed" CPU Tensor or modify it in any ways.

albanD reviewed (changes requested):
I am pretty sure we can prove that this is safe to do. But we should write that down explicitly and add comments to ensure that anyone modifying this code is aware of these facts.

rohan-varma committed:
Update on "Minor optimization to save_on_cpu"


Offload to CPU by using `tensor.to("cpu", non_blocking=True)` instead of allocating an empty tensor and then non-blocking copy to it.

Also, cache `torch.cuda.is_available()` in the constructor.

Overall this provides a 3-4% win on some workloads with large model + large activations that we are testing.

Differential Revision: [D33246189](https://our.internmc.facebook.com/intern/diff/D33246189/)

**NOTE FOR REVIEWERS**: This PR has internal Facebook specific changes or comments, please review them on [Phabricator](https://our.internmc.facebook.com/intern/diff/D33246189/)!

[ghstack-poisoned]

albanD commented on torch/autograd/graph.py:
```diff
@@ -86,7 +86,8 @@ class save_on_cpu(saved_tensors_hooks):
     Args:
         pin_memory (bool): If ``True`` tensors will be saved to CPU pinned memory
-                           during packing and copied to GPU asynchronously during unpacking.
+                           via asynchronous transfer to CPU during packing and
```
nit the sentence is a bit weird with both asynchronous and asyncronously.
Maybe simply: If ``True`` tensors will use pinned memory and perform transfers between the CPU and the GPU asynchronously.

albanD commented on torch/autograd/graph.py:
```diff
@@ -112,21 +113,24 @@ class save_on_cpu(saved_tensors_hooks):
+    # NOTE: when pin_memory=True, save_on_cpu does asynchronous copy when moving
+    # tensor to CPU and GPU. In general, this means that it is the user's job
+    # to do necessary synchronizations, however in this case it is impossible
+    # for the end user to ever have access to this tensor saved by autograd. Any
```
nit: this tensor -> the cpu tensor

albanD commented on torch/autograd/graph.py:
```diff
@@ -112,21 +113,24 @@ class save_on_cpu(saved_tensors_hooks):
+    # tensor to CPU and GPU. In general, this means that it is the user's job
+    # to do necessary synchronizations, however in this case it is impossible
+    # for the end user to ever have access to this tensor saved by autograd. Any
+    # modification to this feature should ensure that this remain the case.
```
nit remains?

albanD commented on torch/autograd/graph.py:
```diff
@@ -112,21 +113,24 @@ class save_on_cpu(saved_tensors_hooks):
         >>> # all intermediary tensors are released (deleted) after the call to backward
 
     """
+    # NOTE: when pin_memory=True, save_on_cpu does asynchronous copy when moving
```
This properly explains why this is safe wrt the user but not why this function is safe.
In particular, we want to mention that we rely on the following facts:
- Any access to the packed Tensor within the SavedVariable is guaranteed to be done through the "unpack" hook.
- The autograd engine is guaranteed to properly restore the current stream that was used during forward before unpacking the SavedVariables for backward.

But there are two points that we are missing I think:
- What is ensuring that when the user accesses this from outside the backward via `grad_fn.saved_foo` that the current stream is properly set? 
- What ensures that the saved cpu tensor is kept alive long enough for the copy back to the gpu to happen?

rohan-varma replied:
> What is ensuring that when the user accesses this from outside the backward via grad_fn.saved_foo that the current stream is properly set?

Do you mean accessing these tensors with `grad_fn.saved_tensors` outside of backward? Is this intended to be a supported feature, as I thought we were assuming this packed tensor should never be accessed by the user (and there's no way to do so)? 

> What ensures that the saved cpu tensor is kept alive long enough for the copy back to the gpu to happen?

I think as you mentioned https://github.com/pytorch/pytorch/blob/3c2ae2b47cc859847e7915a406a7484401cba35e/aten/src/ATen/native/cuda/Copy.cu#L227 should prevent against deallocation happening?

albanD replied:
> Do you mean accessing these tensors with grad_fn.saved_tensors outside of backward? Is this intended to be a supported feature, as I thought we were assuming this packed tensor should never be accessed by the user (and there's no way to do so)?

Yes this is a supported feature. mostly used by visualization tools. But still. 
I think this can be fixed by making sure that the streaming is done properly during unpacking.

> should prevent against deallocation happening?

Yes. We should verify that this is what it is doing and add that to the list of things that we rely on.

> In addition, would there be a difference in the semantics with regard to copy back to GPU?

Yes because now the CPU memory might not be filed when we reach the unpack function. So the unpack function needs to be extra careful that it won't try to read it until it is ready (mostly by ordering properly wrt to the packing's ambient stream I think).
